File: app/src/parts/progress.ts
/** The words and the arithmetic behind the live bar.
 *
 * A run that is working and a run that has hung look the same in a window that shows
 * neither. This turns one step of a run into a line somebody can read at a glance: what
 * it is doing, how long it has been doing it, and how fast the answer is arriving.
 *
 * It is plain because it is tested. The component draws what this returns.
 */

import type { Activity, Step, Waiting } from "../types";

/** What each phase is called, in the words a person would use.
 *
 * The engine reports keys. An unknown key is shown as it came, because a phase the
 * window has no word for is still better than no phase at all.
 */
export const PHASE: Record<string, string> = {
  starting: "starting",
  diff: "reading the change",
  index: "indexing files",
  embed: "embedding code",
  retrieve: "gathering related code",
  rerank: "ranking context",
  scan: "running the scanner",
  outline: "reading the shape of it",
  reading: "reading files",
  asking: "asking the model",
  tool: "running a tool",
  parsing: "reading the answer",
  repairing: "asking for a readable answer",
  verifying: "second opinion",
  saving: "writing findings down",
  posting: "posting the review",
  done: "finished",
};

/** What each job is called. */
export const KIND: Record<string, string> = {
  diff_review: "review",
  audit: "audit",
  pr_review: "pull request",
  verify: "verify",
};

/** Whether the model is still reading the prompt rather than writing an answer.
 *
 * Both happen inside one phase, and on a large prompt the reading is the longer half:
 * a count that is rising while no token has been written yet is the difference between
 * a slow model and a stuck one.
 */
export function reading(step: Step): boolean {
  return step.phase === "asking" && step.tokens === 0 && step.total > 0;
}

export function phaseOf(step: Step): string {
  if (reading(step)) return "reading the prompt";
  return PHASE[step.phase] ?? step.phase;
}

export function kindOf(step: Step): string {
  return KIND[step.kind] ?? step.kind;
}

/** How long since a moment, in the shortest form that is still true.
 *
 * Seconds up to a minute, because that is the range where a person is deciding whether
 * to worry. Both arguments are seconds since the epoch.
 */
export function since(at: number, now: number): string {
  const seconds = Math.max(0, Math.floor(now - at));
  if (seconds < 60) return `${seconds}s`;
  const minutes = Math.floor(seconds / 60);
  if (minutes < 60) return `${minutes}m ${String(seconds % 60).padStart(2, "0")}s`;
  return `${Math.floor(minutes / 60)}h ${String(minutes % 60).padStart(2, "0")}m`;
}

/** How far through a countable phase, from 0 to 1, or null when it cannot be counted. */
export function fraction(step: Step): number | null {
  if (step.total <= 0) return null;
  return Math.min(1, Math.max(0, step.done / step.total));
}

/** How fast the answer is arriving, or "" before there is enough to say.
 *
 * A rate is what says the model is working rather than stuck: a count that is large but
 * still says nothing about whether it moved since the last look.
 */
export function rate(step: Step, now: number): string {
  if (step.tokens <= 0 || step.tokens_started <= 0) return "";
  const seconds = now - step.tokens_started;
  if (seconds < 1) return "";
  return `${(step.tokens / seconds).toFixed(1)}/s`;
}

/** The count beside a phase, when there is one worth showing. */
export function counted(step: Step): string {
  if (step.tokens > 0) return `${step.tokens} tokens`;
  if (reading(step)) return `${step.done}/${step.total} tokens read`;
  if (step.total > 0) return `${step.done}/${step.total}`;
  return "";
}

/** One step as a single line: what, where, how long. */
export function line(step: Step, now: number): string {
  const parts = [phaseOf(step), since(step.phase_started, now)];
  const count = counted(step);
  if (count) parts.splice(1, 0, count);
  if (step.detail) parts.splice(1, 0, step.detail);
  return parts.join(" · ");
}

/** Why a task is being held back, in the words a person would use. */
export const HELD: Record<string, string> = {
  agent_running: "an agent is working in them",
  busy: "they are being worked in",
  machine_in_use: "somebody is at the keyboard",
  model_down: "no model is answering",
  in_flight: "another review of them is running",
};

export function heldFor(reason: string): string {
  return HELD[reason] ?? reason.replace(/_/g, " ");
}

/** The reason the most tasks are waiting for, and how many. */
export function commonest(waiting: Waiting[]): { reason: string; count: number } | null {
  if (waiting.length === 0) return null;
  const counts = new Map<string, number>();
  for (const one of waiting) counts.set(one.reason, (counts.get(one.reason) ?? 0) + 1);
  const [reason, count] = [...counts.entries()].sort((first, second) => second[1] - first[1])[0];
  return { reason, count };
}

/** What the bar says when nothing is in flight.
 *
 * Stopped is a state the user chose, so it is said plainly rather than dressed up as
 * activity. The case that reads as a hang is a queue that is not moving, and it nearly
 * always is not moving for a reason the engine already knows: a coding agent is in the
 * repository, or somebody is at the keyboard. Saying "waiting for a free worker" when
 * every worker is free is the sentence that makes a working rig look broken.
 */
export function resting(activity: Activity | null, now: number): string {
  if (activity === null) return "Connecting";
  if (!activity.ready) return "Starting up";
  if (activity.paused) {
    return activity.pending > 0 ? `Stopped · ${activity.pending} waiting` : "Stopped";
  }
  const waiting = activity.waiting ?? [];
  const held = commonest(waiting);
  if (held !== null) {
    const soonest = Math.min(...waiting.map((one) => one.until));
    const again = soonest > now ? `, retry in ${since(now, soonest)}` : "";
    const rest = activity.pending - waiting.length;
    const others = rest > 0 ? ` · ${rest} queued` : "";
    return `${held.count} held back: ${heldFor(held.reason)}${again}${others}`;
  }
  if (activity.pending > 0) return `${activity.pending} waiting for a free worker`;
  return `Watching · nothing to review${lastly(activity, now)}`;
}

/** What finished last, so a quiet rig still says when it last did something. */
export function lastly(activity: Activity | null, now: number): string {
  const run = activity?.last;
  if (!run || !run.finished_at) return "";
  const at = Date.parse(/(?:Z|[+-]\d\d:\d\d)$/.test(run.finished_at) ? run.finished_at : `${run.finished_at}Z`);
  if (Number.isNaN(at)) return "";
  const found =
    run.status === "ok"
      ? run.finding_count === 0
        ? "nothing found"
        : `${run.finding_count} found`
      : run.status;
  return ` · last ${since(at / 1000, now)} ago, ${found}`;
}

File: engine/src/auger/progress.py
"""Where the work has got to, while it is still running.

A run publishes when it starts and when it finishes, and in between it can spend minutes
inside one step. Without this the window cannot tell work from a hang, so the honest
report is a phase, when the phase started, and how far through it is.

The tracker holds live state only. Nothing here reaches the store or the HTTP layer, and
a job that was given no handle still runs: `nowhere()` answers every call and reports
nothing.
"""

from __future__ import annotations

import itertools
import time
from collections.abc import Callable
from dataclasses import dataclass
from typing import Any

#: How often progress inside one phase is published. A phase change is always published,
#: because it is the part a person is waiting to read. Progress inside one is not: an
#: embedding loop over a few thousand chunks would otherwise fill every subscriber's
#: queue and push the events that matter out of it.
EVERY = 0.5

#: The phases a job reports. The window turns these into words; the engine only has to
#: agree with itself about the keys.
PHASES = (
    "starting",
    "diff",
    "index",
    "embed",
    "retrieve",
    "rerank",
    "scan",
    "outline",
    "reading",
    "asking",
    "tool",
    "parsing",
    "repairing",
    "verifying",
    "saving",
    "posting",
)

_ids = itertools.count(1)


@dataclass
class Step:
    """One run in flight."""

    id: int
    repo: str
    slug: str
    kind: str
    #: Wall clock, because the window subtracts it from the reader's own clock.
    started: float
    phase: str = "starting"
    phase_started: float = 0.0
    detail: str = ""
    #: Progress inside a countable phase. `total` of 0 means the phase cannot be counted,
    #: which is the honest answer for most of them.
    done: int = 0
    total: int = 0
    #: Answer tokens seen so far, while a model is streaming one.
    tokens: int = 0
    #: When the current answer began arriving, so a rate can be worked out.
    tokens_started: float = 0.0
    run: str = ""

    def as_dict(self) -> dict[str, Any]:
        return {
            "id": self.id,
            "repo": self.repo,
            "slug": self.slug,
            "kind": self.kind,
            "started": self.started,
            "phase": self.phase,
            "phase_started": self.phase_started,
            "detail": self.detail,
            "done": self.done,
            "total": self.total,
            "tokens": self.tokens,
            "tokens_started": self.tokens_started,
            "run": self.run,
        }


Publish = Callable[[str, dict[str, Any]], None]


class Watch:
    """One run's progress, as the jobs see it.

    Every method is safe to call at any time and none of them raise, because a report
    about the work must never be able to stop the work.
    """

    def __init__(
        self,
        step: Step,
        publish: Publish | None = None,
        clock: Callable[[], float] = time.time,
    ) -> None:
        self.step = step
        self._publish = publish
        self._clock = clock
        self._said = 0.0

    def phase(self, name: str, detail: str = "", total: int = 0) -> None:
        """Move to a new phase. Always published."""
        now = self._clock()
        self.step.phase = name
        self.step.detail = detail
        self.step.phase_started = now
        self.step.done = 0
        self.step.total = total
        self.step.tokens = 0
        self.step.tokens_started = 0.0
        self._say(now, force=True)

    def advance(self, done: int, total: int | None = None, detail: str | None = None) -> None:
        """How far through a countable phase. Published at a bounded rate."""
        self.step.done = done
        if total is not None:
            self.step.total = total
        if detail is not None:
            self.step.detail = detail
        self._say(self._clock())

    def prefill(self, done: int, total: int) -> None:
        """How much of the prompt the server has read.

        A large prompt spends minutes being read before the first token of the answer
        exists. That is the longest silence in the whole rig, and the phase it happens
        in is the same one the answer arrives in, so it is counted rather than named.
        """
        self.step.done = done
        self.step.total = total
        self._say(self._clock())

    def tokens(self, count: int) -> None:
        """How much of an answer has arrived. Published at a bounded rate."""
        now = self._clock()
        if self.step.tokens_started == 0.0:
            self.step.tokens_started = now
        self.step.tokens = count
        self._say(now)

    def names_run(self, run: str) -> None:
        """The run row this step is writing to, so the window can reach it."""
        self.step.run = run
        self._say(self._clock(), force=True)

    def _say(self, now: float, force: bool = False) -> None:
        if self._publish is None:
            return
        if not force and now - self._said < EVERY:
            return
        self._said = now
        self._publish("run.progress", self.step.as_dict())


class Activity:
    """Every run in flight, newest last."""

    def __init__(
        self, publish: Publish | None = None, clock: Callable[[], float] = time.time
    ) -> None:
        self._publish = publish
        self._clock = clock
        self._live: dict[int, Step] = {}

    def begin(self, repo: str, slug: str, kind: str) -> Watch:
        now = self._clock()
        step = Step(
            id=next(_ids),
            repo=repo,
            slug=slug,
            kind=kind,
            started=now,
            phase_started=now,
        )
        self._live[step.id] = step
        watch = Watch(step, self._publish, self._clock)
        watch._say(now, force=True)
        return watch

    def end(self, watch: Watch) -> None:
        self._live.pop(watch.step.id, None)
        if self._publish is not None:
            self._publish("run.progress", {**watch.step.as_dict(), "phase": "done"})

    def steps(self) -> list[Step]:
        return sorted(self._live.values(), key=lambda step: step.id)


def nowhere() -> Watch:
    """A handle that reports nothing, for a job called without one."""
    return Watch(Step(id=0, repo="", slug="", kind="", started=0.0))
